### PR TITLE
exec: preserve nulls after slicing

### PR DESCRIPTION
**Pull request by asubiotto**
Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Nice find!

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-:-Lbyw5qh6xKCweHGG13I:bdkgtm2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyuWuu5QMXM3_XUtqx:-LbyuWuu5QMXM3_XUtqy:bikpx91) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L259)):*
> ```Go
> 		var nulls []int64
> 		if m.hasNulls {
> 			mod := start % 64
> ```

Not sure how exactly mod for uint64 works, but maybe it's worth making sure that `mod` is non-negative?
___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyuyCF8dTX1jjn2b1a:-LbyuyCF8dTX1jjn2b1b:bosf2wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L263)):*
> ```Go
> 			endIdx := end>>6 + 1
> 			nulls = m.nulls[startIdx:endIdx]
> 			if mod != 0 {
> ```

Again, not sure about this, but do we need to allocate a new underlying array when mod is non-zero?
___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-Lbyv8hg0Vo1pQ4OsVb8:-Lbyv8hh1LIZmlCeRQPX:b-u3yp8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L269)):*
> ```Go
> 					// Bring the first null to the beginning.
> 					nulls[j] = m.nulls[i] >> mod
> 					// And now or the remaining bits with the bits we want to bring over
> ```

[nit]: was briefly confused by the wording, maybe `s/or/bitwise or/` or `s/or/OR/`?
___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyvcVSBCcCG_FB9TT9:-LbyvcVSBCcCG_FB9TTA:bwenunc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L273)):*
> ```Go
> 					nulls[j] |= (m.nulls[i+1] << (64 - mod))
> 				}
> 				// Get the first bit to where we want it for endIdx-1.
> ```

[nit]: bits.

File: pkg/sql/exec/coldata/vec_tmpl.go [hunk 1]
```diff
@@ -254,8 +254,30 @@ func (m *memColumn) Slice(colType types.T, start uint64, end uint64) Vec {
...
+				// If start is not a multiple of 64, we need to shift over the bitmap
+				// to have the first index correspond.
+				for i, j := startIdx, 0; i < endIdx-1; i, j = i+1, j+1 {
+					// Bring the first null to the beginning.
+					nulls[j] = m.nulls[i] >> mod
+					// And now or the remaining bits with the bits we want to bring over
+					// from the next index, note that we handle endIdx-1 separately.
+					nulls[j] |= (m.nulls[i+1] << (64 - mod))
+				}
+				// Get the first bit to where we want it for endIdx-1.
```
Comment (yuzefovich) [inline pkg/sql/exec/coldata/vec_tmpl.go]: Reviewable doesn't like my comments about comments in the code. That empty comment was supposed to be: [nit]: "first bits"

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36637)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyuWuu5QMXM3_XUtqx:-Lc161fT--3VqNzMvLcs:bju8ajn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L259)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Not sure how exactly mod for uint64 works, but maybe it's worth making sure that `mod` is non-negative?
</blockquote></details>

I think it's fine to assume that it won't be given that `start` is a uint64
___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyuyCF8dTX1jjn2b1a:-Lc168AuCPQQMnI8MR8K:brprobj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L263)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Again, not sure about this, but do we need to allocate a new underlying array when mod is non-zero?
</blockquote></details>

Good point. Done, although maybe we could add a flag to specify the desired behavior. cc @jordanlewis, what do you think?
___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-Lbyv8hg0Vo1pQ4OsVb8:-Lc16Qb9A0G6qrAvCIzl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L269)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: was briefly confused by the wording, maybe `s/or/bitwise or/` or `s/or/OR/`?
</blockquote></details>

Done.
___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyqX7KQziGOsE-pCCn-r1-273:-Lc16idz5vdC6tDy9Qej:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L273)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Reviewable doesn't like my comments about comments in the code. That empty comment was supposed to be: [nit]: "first bits"
</blockquote></details>

Done.

**Comment by asubiotto** [comment]
Friendly ping, probably only needs a rubber stamp

**Review by jordanlewis — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-:-LcBecdsExcT4Z1oeiKz:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/exec/coldata/vec_tmpl.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36637#-LbyuyCF8dTX1jjn2b1a:-LcBeLwG8EOY_YF5fneC:bf1e6v5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a8e206ba8717c74402a3e7cf11b277975875e7f/pkg/sql/exec/coldata/vec_tmpl.go#L263)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Good point. Done, although maybe we could add a flag to specify the desired behavior. cc @jordanlewis, what do you think?
</blockquote></details>

This looks good to me, though probably it's unlikely that the fast path will happen too often unless it's common that we slice on 64 multiples...

**Comment by asubiotto** [comment]
TFTRs

bors r=yuzefovich,jordanlewis

**Pull request merged**